contracts/GenArt721CoreV3.sol

```diff
@@ -155,29 +158,66 @@ contract GenArt721CoreV3 is ERC721, Ownable, IGenArt721CoreContractV3 {
         internal
         returns (uint256 _tokenId)
     {
-        uint256 nextTokenId = (_projectId * ONE_MILLION) +
-            projects[_projectId].invocations;
+        // checks & effects
+        // increment project's invocations, then move to memory to avoid SLOAD
+        uint256 _invocationsAfter = ++projects[_projectId].invocations;
+        uint256 _invocationsBefore = _invocationsAfter - 1;
+        uint256 thisTokenId = (_projectId * ONE_MILLION) + _invocationsBefore;
 
-        projects[_projectId].invocations++;
+        // mark project as completed if hit max invocations
+        if (_invocationsAfter == projects[_projectId].maxInvocations) {
+            _completeProject(_projectId);
+        }
 
         bytes32 tokenHash = keccak256(
             abi.encodePacked(
-                projects[_projectId].invocations,
+                thisTokenId,
                 blockhash(block.number - 1),
                 randomizerContract.returnValue()
             )
         );
 
-        tokenIdToHash[nextTokenId] = tokenHash;
+        tokenIdToHash[thisTokenId] = tokenHash;
 
-        _mint(_to, nextTokenId);
+        // interactions
+        _mint(_to, thisTokenId);
 
         // Do not need to also log `projectId` in event, as the `projectId` for
         // a given token can be derived from the `tokenId` with:
         //   projectId = tokenId / 1_000_000
-        emit Mint(_to, nextTokenId);
+        emit Mint(_to, thisTokenId);
 
-        return nextTokenId;
+        return thisTokenId;
+    }
+
+    /**
+     * @notice Returns whether a project is locked.
+     * Projects automatically lock four weeks after they are completed.
+     * @param _projectId Project ID to check.
+     */
+    function projectLocked(uint256 _projectId) public view returns (bool) {
+        return
+            projectCompleted(_projectId) &&
+            (block.timestamp - projects[_projectId].completedTimestamp >
+                FOUR_WEEKS_IN_SECONDS);
+    }
+
+    /**
+     * @notice Returns whether a project is completed.
+     * A project is considered completed when its invocations equals its max
+     * invocations.
+     * @param _projectId Project ID to check.
+     */
+    function projectCompleted(uint256 _projectId) public view returns (bool) {
+        return projects[_projectId].completedTimestamp != 0;
```

Comment:
Is `!=` two ops, an `EQ` and a `NOT`? I can't tell from https://ethereum.org/en/developers/docs/evm/opcodes/

If so, I wonder if we want to just make this `projectOpen` rather than `projectCompleted` to avoid the extra `!` here, and then make `projectLocked` to be `projectUnlocked` as well so there is less negating within the function and with its usage (because we actually check for unlocked status, not locked status).

Reply:
I quite like this suggestion because projectUnlocked is actually much easier to prove (can short-circuit when completedTimestamp == 0). It also seems more intuitive (less negating).

I implemented it, and reduced the public function clutter by making the functions internal, since project state is available on `projectStateData()`. Note that I am not exposing complete timestamp because I assume having that available in our subgraph is sufficient. We could expose it in projectStateData as another return value, but we would have to educate the user what it means when the value returned is zero. I also return `locked` in the projectStateData function because in V1, we return `bool locked` in the view function `projectScriptInfo`.

Reply:
> I implemented it, and reduced the public function clutter by narrowing everything down to a single internal function (since a view of project state is available on projectStateData()). Note that I am not exposing complete timestamp because I assume having that available in our subgraph is sufficient. We could expose it in projectStateData as another return value, but we would have to educate the user what it means when the value returned is zero. I also return locked in the projectStateData function because in V1, we return locked (not unlocked) in the V1 view function projectScriptInfo.

This rationale sounds good to me–agreed.